core/src/components/cat-scrollable/cat-scrollable.tsx

```diff
@@ -0,0 +1,156 @@
+
+  /** Flags to enable/disable scroll shadowX. */
+  @Prop()
+  shadowX = true;
+
+  /** Flags to enable/disable scroll shadowY. */
+  @Prop()
+  shadowY = true;
+
+  /** Flags to enable/disable overflowX. */
+  @Prop()
+  overflowX = true;
+
+  /** Flags to enable/disable overflowY. */
+  @Prop()
+  overflowY = true;
+
+  /** Flag to enable/disable overscroll behavior. */
+  @Prop()
+  overscroll = true;
+
+  /**
+   * Flag to fire an initial event after content initialization.
+   */
+  @Prop() scrolledInit = true;
+
+  /**
+   * Buffer to be used to calculate the scroll distance.
+   */
+  @Prop()
+  scrolledBuffer = 0;
+
+  @Event() scrolledBottom!: EventEmitter<void>;
+  @Event() scrolledTop!: EventEmitter<void>;
+  @Event() scrolledLeft!: EventEmitter<void>;
+  @Event() scrolledRight!: EventEmitter<void>;
+
+  componentDidRender() {
+    if (this.scrollElement) {
+      this.scrolled = fromEvent(this.scrollElement, 'scroll')
+        .pipe(takeUntil(this.destroyed));
+    }
+    this.attachEmitter('left', this.scrolledLeft, this.scrolledBuffer);
+    this.attachEmitter('right', this.scrolledRight, this.scrolledBuffer);
+    this.attachEmitter('bottom', this.scrolledBottom, this.scrolledBuffer);
+    this.attachEmitter('top', this.scrolledTop, this.scrolledBuffer);
+    merge(this.init, this.scrolled)
+      .pipe(map(() => ({
+          top: this.getScrollOffset('top') > 0,
+          left: this.getScrollOffset('left') > 0,
+          right: this.getScrollOffset('right') > 0,
+          bottom: this.getScrollOffset('bottom') > 0
+        })),
+        distinctUntilChanged(),
+        takeUntil(this.destroyed)
+      )
+      .subscribe(({ top, left, right, bottom }) => {
+        this.toggleClass('cat-scrollable-top', top);
+        this.toggleClass('cat-scrollable-left', left);
+        this.toggleClass('cat-scrollable-right', right);
+        this.toggleClass('cat-scrollable-bottom', bottom);
+      });
+  }
+
+  componentDidLoad() {
+    this.checkInit(this.scrolledInit);
+  }
+
+  disconnectedCallback() {
+    this.init.complete();
+    this.destroyed.next();
+    this.destroyed.complete();
+  }
+
+  render() {
+    return (
+      <Host class={{
+        'scroll-x': this.overflowX,
+        'scroll-y': this.overflowY,
+        'no-overscroll': !this.overscroll
+      }}>
+        <div class='shadow-wrapper'
+             ref={el => (this.scrollShadowElement = el)}>
+          {this.shadowY && <div class='shadow-top'></div>}
+          {this.shadowX && <div class='shadow-left'></div>}
+          {this.shadowX && <div class='shadow-right'></div>}
+          {this.shadowY && <div class='shadow-bottom'></div>}
+        </div>
+        <div
+          ref={el => (this.scrollElement = el)}
+          class='scrollable-content'>
+          <slot></slot>
+        </div>
+      </Host>
+    );
+  }
+
+  private attachEmitter(from: 'top' | 'left' | 'right' | 'bottom', emitter: EventEmitter<void>, buffer: number) {
```

Comment:
no return types for private methods.

Reply:
Also, please avoid the use of the internal type/interfaces, move `'top' | 'left' | 'right' | 'bottom'` separately.